OfflineUX.md

```diff
@@ -0,0 +1,138 @@
+
+1. **API Request:** We send a request to the API to ensure the change is saved in the database. 
+
+2. **UI Changes:** The chat should go to the top of the list with the other pinned chats, and the pin button should look darker than it did before. This is visual feedback that clicking the pin button worked.
+
+If the user is offline, we don't need to wait for the API request to finish before doing all that visual stuff because this particular API request has almost no way of failing and we know what the server will return in advance. That means we can safely assume that when we send the command, it will succeed and that's why we let the user continue using the app as if the action succeeded.
+</div>
+
+The example we just looked at is nice and simple, but some actions should use this apporach (example: requesting money from another user). For these types of actions, we can’t simply proceed as if the request already finished. Here are some reasons why:
+
+1. In some instances, we simply wouldn't know _how_ to proceed because of a lack of information (often the server returns data that we wouldn't be able to guess the content of). 
+
+2. In other cases, we may be able to guess what a successful request would look like, but we don't want to misguide the user into believing their action was completed. For example, we don't want the user to believe that a financial transaction has been made when it actually hasn't. 
+
+To handle problems like this, we have developed offline patterns and guidance on when to use them. These are called the Offline UX Patterns, which are explained below.
+
+### Descriptions of the UX Patterns
+
+# None - No Offline Behavior
+
+ There’s no specific UI for this case. The feature either looks totally normal and works as expected (because it doesn’t need the server to function) or the feature looks like it did whenever connection was lost. 
+
+Used when…
+ - there is no interaction with the server in any way
+ - or data is READ from the server and does not need to show up-to-date data. The user will see stale data until the new data is put into Onyx and then the view updates to show the new data. 
+
+# A - Optimistic Without Feedback Pattern
+
+This is the pattern where we queue the request to be sent when the user is online and we continue as if the request succeeded. 
+
+Used when…
+ - the user should be given instant feedback and
+ - there is no error handling in the server and
+ - the user does not need to know when the change is done on the server in the background
+
+How to implement: Use API.write() to implement this pattern. 
+
+# B - Optimistic WITH Feedback Pattern
+This pattern queues the API request, but also makes sure that the user is aware that the request hasn’t been sent yet **when the user is offline**. When the user is online, the feature should just look like it succeeds immediately (we dont want the offline UI to flicker on and off when the user is online).
```

Comment:
NAB: I think we technically show a feature (i.e. `Report_AddComment`) in a "pending" state when the request is completed, no? Because the request can be sent and we won't change the UI of the feature until we get the server response indicating that the feature shouldn't be in a pending state anymore (i.e. loading = false).

> `When the user is online, the feature should just look like it succeeds immediately`

I think it doesn't apply perfectly to the `Report_AddComment` example, right? Where the UI of the comment will have a light gray color and will be darkened when the server response pushes the value `loading: false` into Onyx. What if we change it to:

```suggestion
This pattern queues the API request, but also makes sure that the user is aware that the request hasn’t been completed yet **when the user is offline**. The feature will have a slightly different UI to indicate that it's in a pending state (i.e. the Report comments are grayed out when the request is not completed). When the request is completed, the feature should just look in its normal state.
```

Reply:
No, only if we are offline. This will not happen if we are online. If we are online we just assume the action will be created without any pending state.